scripts/keyframes.py

```diff
@@ -0,0 +1,422 @@
+#!/usr/bin/env python3
+"""Given a gfxreconstruct command file, identify "interesting"
+keyframes based on distinct maximal function sets used in a frame.  Keyframes
+selected in this way tend to exercise more distinct parts of the trace,
+as many frames will use the same "footprint" of functions as other frames;
+by collapsing to maximal function sets, keyframes can be selected that more
+thoroughly exercise the graphics API.
+
+The selected keyframes can be further refined through the use of filter
+operations.
+
+To use on a trace captured by gfxreconstruct, the trace must be converted
+to an ASCII command file containing one function or method name per line, e.g.:
+
+    ...
+    vkDestroyImageView
+    vkBeginCommandBuffer
+    vkCmdPipelineBarrier
+    vkCmdWriteTimestamp
+    vkAcquireNextImageKHR
+    vkCmdPipelineBarrier
+    vkCmdBeginRenderPass
+    vkCmdBindPipeline
+    vkCmdBindDescriptorSets
+    vkCmdSetViewport
+    vkCmdBindIndexBuffer
+    vkCmdDrawIndexed
+    vkCmdPipelineBarrier
+    vkCmdWriteTimestamp
+    vkCmdEndRenderPass
+    vkCmdResetQueryPool
+    vkEndCommandBuffer
+    vkFlushMappedMemoryRanges
+    vkResetFences
+    vkQueueSubmit
+    vkQueuePresentKHR
+    ...
+
+The command file must include frame marker functions (e.g. vkQueuePresentKHR
+for Vulkan); if it does not, no frame boundaries can be established, and keyframes
+cannot be calculated.
+
+The gfxrecon-convert tool can be used to create such a file from a trace
+file, using an editor or a script to extract only the function or method
+name associated with each captured call.
+
+Then using this script will yield the keyframes, e.g.:
+
+    $ python3 keyframes.py commandfile.txt
+    INFO: processing the command file commandfile.txt...
+    INFO: sorting 70 function sets by cardinality...
+    INFO: collapsing into maximal distinct sets...
+    INFO: filtering 10 keyframes from 22 maximal sets...
+    INFO: Keyframes in priority order: 0,1908,1051,1910,2086,4280,5329,1616,2088,4281
+    INFO: Keyframes in sorted order:   0,1051,1616,1908,1910,2086,2088,4280,4281,5329
+
+This script can also create a command_counts.json file in the same directory, and
+create stub reference images, each of which is helpful in trace processing.
```

Comment:
It isn't clear to me what a stub reference image is or why it is useful.

Reply:
Added more commentary.